### ISSUE TITLE
JsonSerializer.Parse<DataTable> failed

### ISSUE THREAD

**Issue by xcaptain** [dotnet/corefx#38712]
Convert an object into json string and then convert into DataTable is an easy way to generate `DataTable`, but this approach fails in `dotnet core 3 preview5`

I have created a working dotnet core 2 code snippet at [https://dotnetfiddle.net/cD6NFD](https://dotnetfiddle.net/cD6NFD)

dotnetfiddle doesn't support dotnet core 3 yet, so my dotnet core 3 code example see below:

```csharp
using System.Text.Json.Serialization;

public class UserDetails
{
    public string Id { get; set; }
    public string Name { get; set; }
    public string City { get; set; }
    public string Country { get; set; }
}

List<UserDetails> persons = new List<UserDetails>()
{
    new UserDetails() {Id="1001", Name="ABCD", City ="City1", Country="USA"},
    new UserDetails() {Id="1002", Name="PQRS", City ="City2", Country="INDIA"},
    new UserDetails() {Id="1003", Name="XYZZ", City ="City3", Country="CHINA"},
    new UserDetails() {Id="1004", Name="LMNO", City ="City4", Country="UK"},
};
var table = JsonSerializer.Parse<DataTable>(JsonSerializer.ToString(persons));
```

below are the error log:
```
CLR/System.Text.Json.JsonReaderException 'The JSON value from [System.Data.DataTable] could not be converted to System.Object.'
```

**Comment by ahsonkhan** [comment]
`DataTable` and the related types are not currently supported by the `JsonSerializer` (for 3.0).

cc @steveharter

**Comment by xcaptain** [comment]
Thanks, because NewtonSoft.Json support this feature, would be awesome if `System.Text.Json` support this.

**Comment by AlonCG** [comment]
We ran into this issues as well, and being stubborn (& lazy ... since I did not read all the docs first) ... I decided to trudge through and forcefully and painstakingly remove JSON.Net from our 2.2 to 3.1 migration. In the end we created a `JsonConverter<DataTable>`. 

I have **not tested** it extensively and I am sure there are dozens of other **better** examples out there, but this was my first attempt and is currently working in our situation. Maybe this can be a stepping stone for others instead of just (1) go back to Json.Net or (2) wait for Net 5! 

I'm sure I will refactor this a dozen times before it goes to production but constructive comments are always helpful. If it is not useful (or lacking in any area), then please disregard this post (or admins can delete it!). 
<pre>
public class DataTableJsonConverter : JsonConverter<DataTable>
{
	public override DataTable Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
	{
		...
	}

	public override void Write(Utf8JsonWriter writer, DataTable value, JsonSerializerOptions options)
	{
		CustomDataTableSerializer.WriteDataTable(writer, value, options);
	}
}
</pre>
<pre>
public static class CustomDataTableSerializer
{
	public static void WriteDataTable(Utf8JsonWriter jsonWriter, DataTable source, JsonSerializerOptions options)
	{
		jsonWriter.WriteStartArray();
		foreach (DataRow dr in source.Rows) {
			jsonWriter.WriteStartObject();
			foreach (DataColumn col in source.Columns) {
				var key = col.ColumnName.Trim();
				var valueString = dr[col].ToString();
				switch (col.DataType.FullName) {
					case "System.Guid":
						jsonWriter.WriteString(JsonEncodedText.Encode(key), valueString);
						break;
					case "System.Char":
					case "System.String":
						jsonWriter.WriteString(JsonEncodedText.Encode(key), JsonEncodedText.Encode(valueString));
						break;
					case "System.Boolean":
						Boolean.TryParse(valueString, out bool boolValue);
						jsonWriter.WriteBoolean(JsonEncodedText.Encode(key), boolValue);
						break;
					case "System.DateTime":
						var dateValue = DateTime.Parse(valueString);
						jsonWriter.WriteString(JsonEncodedText.Encode(key), dateValue);
						break;
					case "System.TimeSpan":
						var timeSpanValue = TimeSpan.Parse(valueString);
						jsonWriter.WriteString(JsonEncodedText.Encode(key), timeSpanValue.ToString());
						break;
					case "System.Byte":
					case "System.SByte":
					case "System.Decimal":
					case "System.Double":
					case "System.Single":
					case "System.Int16":
					case "System.Int32":
					case "System.Int64":
					case "System.UInt16":
					case "System.UInt32":
					case "System.UInt64":
						if (Int64.TryParse(valueString, out long intValue)) {
							jsonWriter.WriteNumber(JsonEncodedText.Encode(key), intValue);
						} else {
							Double.TryParse(valueString, out double doubleValue);
							jsonWriter.WriteNumber(JsonEncodedText.Encode(key), doubleValue);
						}
						break;
					default:
						jsonWriter.WriteString(JsonEncodedText.Encode(key), JsonEncodedText.Encode(valueString));
						break;
				}
			}
			jsonWriter.WriteEndObject();
		}
		jsonWriter.WriteEndArray();
	}
}
</pre>
I reuse `WriteDataTable` in multiple spots, hence the reason it is separate. Once we get this into our staging environment, I'll most probably find missing edge-cases and possibly more efficient ways to write this out. 

Happy coding!

**Comment by ahsonkhan** [comment]
From a quick glance, I would recommend one adjustment. Remove **all** the calls to `JsonEncodedText.Encode` and pass the strings directly to the writer. Unless you are doing custom encoding specific to a particular key/value pair, it is just extra overhead and making things slower (+ you are writing more code). `JsonEncodedText` is mainly useful for known/constant string literals you want to encode up front and store in statics to re-use. And if you want to globally change the encoding behavior, change the `Encoder` setting on `JsonWriterOptions` instead.

For example:
```diff
- jsonWriter.WriteString(JsonEncodedText.Encode(key), JsonEncodedText.Encode(valueString));
+ jsonWriter.WriteString(key, valueString);
```

**Comment by AlonCG** [comment]
@ahsonkhan Yeah, that is simple. For some reason early on I thought it was required, so just copied and pasted it there! Thanks!

**Comment by etghanibahman** [comment]
@AlonCG The main problem is when you are going to read serialized data and turn it to dataset.
In this case you do not know what is the type of column, so you do not know you can use which function to retrieve data.

**Comment by AlonCG** [comment]
@etghanibahman, yeah ... I basically just do the reverse, hence the "..." in the *Read* function. 

This needs some work, since I don't like the double enumeration and could use some optimization (and I am not sure it is even a good idea, but I needed reusable bits)... but again, this could just be used as a stepping stone for others that are stuck and need a bit of a boost?

````
public class DataTableJsonConverter : JsonConverter<DataTable>
{
    public override DataTable Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
    {
        using var jsonDoc = JsonDocument.ParseValue(ref reader);
        var rootElement = jsonDoc.RootElement;
        var dataTable = rootElement.JsonElementToDataTable();
        return dataTable;
    }

    public override void Write(Utf8JsonWriter writer, DataTable value, JsonSerializerOptions options)
    {
        ... (see above)
    }
}
````
JsonElementToDataTable Extension and support functions (could use some work for sure!)
````
public static DataTable JsonElementToDataTable(this JsonElement dataRoot)
{
    var dataTable = new DataTable();
    var firstPass = true;
    foreach (var element in dataRoot.EnumerateArray()) {
        if (firstPass) {
            foreach (var col in element.EnumerateObject()) {
                var colValue = col.Value;
                dataTable.Columns.Add(new DataColumn(col.Name, colValue.ValueKind.ValueKindToType(colValue.ToString())));
            }
            firstPass = false;
        }
        var row = dataTable.NewRow();
        foreach (var col in element.EnumerateObject()) {
            row[col.Name] = col.Value.JsonElementToTypedValue();
        }
        dataTable.Rows.Add(row);
    }
    return dataTable;
}

private static Type ValueKindToType(this JsonValueKind valueKind, string value)
{
    switch (valueKind) {
        case JsonValueKind.String:      // 3
            return typeof(System.String);
        case JsonValueKind.Number:      // 4    
            if (Int64.TryParse(value, out var intValue)) {
                return typeof(System.Int64);
            } else {
                return typeof(System.Double);
            }
        case JsonValueKind.True:        // 5
        case JsonValueKind.False:       // 6
            return typeof(System.Boolean);
        case JsonValueKind.Undefined:   // 0
            return null;
        case JsonValueKind.Object:      // 1 
            return typeof(System.Object);
        case JsonValueKind.Array:       // 2
            return typeof(System.Array);
        case JsonValueKind.Null:        // 7
            return null;
        default:
            return typeof(System.Object);
    }
}

private static object JsonElementToTypedValue(this JsonElement jsonElement)
{
    switch (jsonElement.ValueKind) {
        case JsonValueKind.Object:      // 1  (these need special handling)?
        case JsonValueKind.Array:       // 2
        case JsonValueKind.String:      // 3
            if (jsonElement.TryGetGuid(out Guid guidValue)) {
                return guidValue;
            } else {
                if (jsonElement.TryGetDateTime(out DateTime datetime)) {
                    // If an offset was provided, use DateTimeOffset.
                    if (datetime.Kind == DateTimeKind.Local) {
                        if (jsonElement.TryGetDateTimeOffset(out DateTimeOffset datetimeOffset)) {
                            return datetimeOffset;
                        }
                    }
                    return datetime;
                }
                return jsonElement.ToString();
            }
        case JsonValueKind.Number:      // 4    
            if (jsonElement.TryGetInt64(out long longValue)) {
                return longValue;
            } else {
                return jsonElement.GetDouble();
            }
        case JsonValueKind.True:        // 5
        case JsonValueKind.False:       // 6
            return jsonElement.GetBoolean();
        case JsonValueKind.Undefined:   // 0
        case JsonValueKind.Null:        // 7
            return null;
        default:
            return jsonElement.ToString();
    }
}
````
While this works for us, *we have not done extensive testing* on it, at all. It was just implemented to get around the issue of not having serialization for DataTables. So instead of waiting for Core 5 or going back to Json.Net (hindsight being 20/20), this is what we got.

**Comment by etghanibahman** [comment]
Thank you, it works really well in my case.
